File: rewrite-java/src/main/java/org/openrewrite/java/marker/TrailingComma.java
/*
 * Copyright 2023 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.marker;

import lombok.Value;
import lombok.With;
import org.openrewrite.java.internal.rpc.JavaReceiver;
import org.openrewrite.java.internal.rpc.JavaSender;
import org.openrewrite.java.tree.Space;
import org.openrewrite.marker.Marker;
import org.openrewrite.rpc.RpcCodec;
import org.openrewrite.rpc.RpcReceiveQueue;
import org.openrewrite.rpc.RpcSendQueue;

import java.util.UUID;

@Value
@With
public class TrailingComma implements Marker, RpcCodec<TrailingComma> {
    UUID id;
    Space suffix;

    @Override
    public void rpcSend(TrailingComma after, RpcSendQueue q) {
        q.getAndSend(after, Marker::getId);
        q.getAndSend(after, TrailingComma::getSuffix, space -> new JavaSender().visitSpace(space, q));
    }

    @Override
    public TrailingComma rpcReceive(TrailingComma before, RpcReceiveQueue q) {
        return before
                .withId(q.receiveAndGet(before.getId(), UUID::fromString))
                .withSuffix(q.receive(before.getSuffix(), space -> new JavaReceiver().visitSpace(space, q)));
    }
}

File: rewrite-csharp/csharp/OpenRewrite/Core/TrailingComma.cs
/*
 * Copyright 2026 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using OpenRewrite.Core.Rpc;
using OpenRewrite.Java.Rpc;

namespace OpenRewrite.Core;

/// <summary>
/// Marker indicating a trailing comma after the last element in a separated list
/// (enum members, initializer expressions, switch arms, etc.).
/// The <see cref="Suffix"/> captures the whitespace/comments between the trailing
/// comma and the closing delimiter (e.g., '}').
/// </summary>
public sealed class TrailingComma(Guid id, Space suffix) : Marker, IRpcCodec<TrailingComma>
{
    public Guid Id { get; } = id;
    public Space Suffix { get; } = suffix;

    public TrailingComma WithId(Guid id) => id == Id ? this : new(id, Suffix);
    public TrailingComma WithSuffix(Space suffix) => ReferenceEquals(suffix, Suffix) ? this : new(Id, suffix);

    public void RpcSend(TrailingComma after, RpcSendQueue q)
    {
        q.GetAndSend(after, m => m.Id);
        q.GetAndSend(after, m => m.Suffix, s => new JavaSender().VisitSpace(s, q));
    }

    public TrailingComma RpcReceive(TrailingComma before, RpcReceiveQueue q) =>
        before
            .WithId(q.ReceiveAndGet<Guid, string>(before.Id, Guid.Parse))
            .WithSuffix(q.Receive(before.Suffix, s => new JavaReceiver().VisitSpace(s, q)));
}
